src/Ztm.WebApi/TransactionConfirmationWatcher.cs

```diff
@@ -0,0 +1,354 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Hosting;
+using NBitcoin;
+using Ztm.Threading;
+using Ztm.Zcoin.Synchronization;
+using Ztm.Zcoin.Watching;
+
+namespace Ztm.WebApi
+{
+    using WatchRepository = ITransactionConfirmationWatchRepository<TransactionConfirmationCallbackResult>;
+    using ConfirmContext = TransactionConfirmationWatch<TransactionConfirmationCallbackResult>;
+
+    public sealed class TransactionConfirmationWatcher : IHostedService, IBlockListener, ITransactionConfirmationWatcherHandler<ConfirmContext>
+    {
+        readonly Ztm.Zcoin.Watching.TransactionConfirmationWatcher<ConfirmContext> watcher;
+
+        // Providers
+        readonly ICallbackRepository callbackRepository;
+        readonly ITransactionConfirmationWatchRepository<TransactionConfirmationCallbackResult> watchRepository;
+        readonly ICallbackExecuter callbackExecuter;
+
+        // State recorders
+        readonly ReaderWriterLockSlim timerLock;
+        readonly Dictionary<uint256, Dictionary<Guid, Tuple<Ztm.Threading.Timer, ConfirmContext>>> timers;
+
+        readonly ConcurrentDictionary<Guid, TransactionWatch<ConfirmContext>> watches;
+
+        public TransactionConfirmationWatcher(
+            ICallbackRepository callbackRepository,
+            WatchRepository watchRepository,
+            IBlocksStorage blocks,
+            ICallbackExecuter callbackExecuter)
+        {
+            if (callbackRepository == null)
+            {
+                throw new ArgumentNullException(nameof(callbackRepository));
+            }
+
+            if (watchRepository == null)
+            {
+                throw new ArgumentNullException(nameof(watchRepository));
+            }
+
+            if (blocks == null)
+            {
+                throw new ArgumentNullException(nameof(blocks));
+            }
+
+            if (callbackExecuter == null)
+            {
+                throw new ArgumentNullException(nameof(callbackExecuter));
+            }
+
+            this.callbackRepository = callbackRepository;
+            this.watchRepository = watchRepository;
+            this.callbackExecuter = callbackExecuter;
+
+            this.watcher = new Zcoin.Watching.TransactionConfirmationWatcher<ConfirmContext>
+            (
+                this,
+                blocks
+            );
+
+            timerLock = new ReaderWriterLockSlim(LockRecursionPolicy.SupportsRecursion);
+            timers = new Dictionary<uint256, Dictionary<Guid, Tuple<Threading.Timer, ConfirmContext>>>();
+            watches = new ConcurrentDictionary<Guid, TransactionWatch<ConfirmContext>>();
+        }
+
+        public async Task Initialize(CancellationToken cancellationToken)
```

Comment:
We should not expose this one due to it duplicated with `StartAsync`. My idea is we can merge this into `StartAsync` due to no benefits to separate it.